sdks/python/apache_beam/coders/coder_impl.py

```diff
@@ -74,53 +86,66 @@
   else:
     is_compiled = False
     fits_in_64_bits = lambda x: -(1 << 63) <= x <= (1 << 63) - 1
+
 # pylint: enable=wrong-import-order, wrong-import-position, ungrouped-imports
 
 
 _TIME_SHIFT = 1 << 63
 MIN_TIMESTAMP_micros = MIN_TIMESTAMP.micros
 MAX_TIMESTAMP_micros = MAX_TIMESTAMP.micros
 
+IterableStateReader = Callable[[bytes, 'CoderImpl'], Iterable]
+IterableStateWriter = Callable[[Iterable, 'CoderImpl'], bytes]
+Observables = List[Tuple[observable.ObservableMixin, 'CoderImpl']]
 
 class CoderImpl(object):
   """For internal use only; no backwards-compatibility guarantees."""
 
   def encode_to_stream(self, value, stream, nested):
+    # type: (Any, create_OutputStream, bool) -> None
```

Comment:
My intent was that mypy would use the slow_stream implementation of these, which _are_ types.  But I realized when looking at this that `create_InputStream` is being treated as `Any` because mypy can't inspect the `stream` c-extension module (mypy always uses the first import location of an object, unless `TYPE_CHECKING` is used). 

The simple/obvious approach works, but is pretty verbose:

```python
if TYPE_CHECKING:
  from apache_beam.transforms.window import IntervalWindow
  from .slow_stream import InputStream as create_InputStream
  from .slow_stream import OutputStream as create_OutputStream
  from .slow_stream import ByteCountingOutputStream
  from .slow_stream import get_varint_size
else:
  # pylint: disable=wrong-import-order, wrong-import-position, ungrouped-imports
  try:
    from .stream import InputStream as create_InputStream
    from .stream import OutputStream as create_OutputStream
    from .stream import ByteCountingOutputStream
    from .stream import get_varint_size
    # Make it possible to import create_InputStream and other cdef-classes
    # from apache_beam.coders.coder_impl when Cython codepath is used.
    globals()['create_InputStream'] = create_InputStream
    globals()['create_OutputStream'] = create_OutputStream
    globals()['ByteCountingOutputStream'] = ByteCountingOutputStream
  except ImportError:
    from .slow_stream import InputStream as create_InputStream
    from .slow_stream import OutputStream as create_OutputStream
    from .slow_stream import ByteCountingOutputStream
    from .slow_stream import get_varint_size
    if False:  # pylint: disable=using-constant-test
      # This clause is interpreted by the compiler.
      from cython import compiled as is_compiled
    else:
      is_compiled = False
      fits_in_64_bits = lambda x: -(1 << 63) <= x <= (1 << 63) - 1
```

Here's another approach (still verbose but less redundant):

```python
try:
  from . import stream
except ImportError:
  SLOW_STREAM = True
else:
  SLOW_STREAM = False

if TYPE_CHECKING or SLOW_STREAM:
  from .slow_stream import InputStream as create_InputStream
  from .slow_stream import OutputStream as create_OutputStream
  from .slow_stream import ByteCountingOutputStream
  from .slow_stream import get_varint_size

  if False:  # pylint: disable=using-constant-test
    # This clause is interpreted by the compiler.
    from cython import compiled as is_compiled
  else:
    is_compiled = False
    fits_in_64_bits = lambda x: -(1 << 63) <= x <= (1 << 63) - 1

else:
  # pylint: disable=wrong-import-order, wrong-import-position, ungrouped-imports
  from .stream import InputStream as create_InputStream
  from .stream import OutputStream as create_OutputStream
  from .stream import ByteCountingOutputStream
  from .stream import get_varint_size
  # Make it possible to import create_InputStream and other cdef-classes
  # from apache_beam.coders.coder_impl when Cython codepath is used.
  globals()['create_InputStream'] = create_InputStream
  globals()['create_OutputStream'] = create_OutputStream
  globals()['ByteCountingOutputStream'] = ByteCountingOutputStream
  # pylint: enable=wrong-import-order, wrong-import-position, ungrouped-imports
```

Opinion?

Reply:
Those look good, with a slight preference for the second.

Reply:
roger.  I'm building up a commit that addresses all of the notes, so I'll leave this unresolved until that's ready.